PopPUNK/sketchlib.py

```diff
@@ -225,35 +225,65 @@ def joinDBs(db1, db2, output):
             Prefix for db2
         output (str)
             Prefix for joined output
+        update_random (dict)
+            Whether to re-calculate the random object. May contain
+            control arguments strand_preserved and threads (see :func:`addRandom`)
     """
-    join_name = output + "/" + os.path.basename(output) + ".h5"
+    join_prefix = output + "/" + os.path.basename(output)
     db1_name = db1 + "/" + os.path.basename(db1) + ".h5"
     db2_name = db2 + "/" + os.path.basename(db2) + ".h5"
 
     hdf1 = h5py.File(db1_name, 'r')
     hdf2 = h5py.File(db2_name, 'r')
-    hdf_join = h5py.File(join_name + ".tmp", 'w') # add .tmp in case join_name exists
+    hdf_join = h5py.File(join_prefix + ".tmp.h5", 'w') # add .tmp in case join_name exists
 
     # Can only copy into new group, so for second file these are appended one at a time
     try:
         hdf1.copy('sketches', hdf_join)
-        if 'random' in hdf1:
-            hdf1.copy('random', hdf_join)
+
         join_grp = hdf_join['sketches']
         read_grp = hdf2['sketches']
         for dataset in read_grp:
             join_grp.copy(read_grp[dataset], dataset)
 
+        # Copy or update random matches
+        if update_random is not None:
+            threads = 1
+            strand_preserved = False
+            if isinstance(update_random, dict):
+                if "threads" in update_random:
+                    threads = update_random["threads"]
+                if "strand_preserved" in update_random:
+                    strand_preserved = update_random["strand_preserved"]
+
+            sequence_names = list(hdf_join['sketches'].keys())
+            kmer_size = hdf_join['sketches/' + sequence_names[0]].attrs['kmers']
+
+            # Need to close before adding random
+            hdf_join.close()
+            if len(sequence_names) > 2:
+                sys.stderr.write("Updating random match chances")
+                pp_sketchlib.addRandom(join_prefix + ".tmp",
+                                       sequence_names,
+                                       kmer_size,
+                                       not strand_preserved,
+                                       threads)
```

Comment:
Does this need checking against the number of groups used for the random sequence generation?

Reply:
I don't think so – it regenerates the whole thing from scratch

Reply:
I couldn't remember - I know it was 10 groups by default, which caused problems with <10 sequences.

Reply:
Default clusters is now 3, and we've got the check above here, so think it should be ok